File: 백준/Gold/1106. 호텔/호텔.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int answer = Integer.MAX_VALUE;
        int C = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());

        List<int[]> info = new ArrayList<>();
        int[] dp = new int[C+101];
        Arrays.fill(dp, Integer.MAX_VALUE);

        for(int i=0; i<N; ++i)
        {
            st = new StringTokenizer(br.readLine());
            int cost = Integer.parseInt(st.nextToken());
            int profit = Integer.parseInt(st.nextToken());
            info.add(new int[]{cost, profit});
            dp[profit] = Math.min(dp[profit], cost);
        }

        for(int p=0; p<C+101; ++p)
        {
            for(int i=0; i<N; ++i)
            {
                int cost = info.get(i)[0];
                int profit = info.get(i)[1];

                if(p < profit) continue;

                if (dp[p - profit] != Integer.MAX_VALUE) {
                    dp[p] = Math.min(dp[p], dp[p - profit] + cost);
               }
            }

            if (p >= C)
                answer = Math.min(answer, dp[p]);
        }

        System.out.print(answer);

    } // main
}

File: 백준/Gold/1106. 호텔/호텔.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main()
{
    int C, N, answer = 1e9;
    int cost, value;
    cin >> C >> N;
    vector<int> dp(C+101, 1e9);
    dp[0] = 0;

    for(int i=0; i<N; ++i)
    {
        cin >> cost >> value; // cost원 들이면 인구수 value 만큼 얻을 수 있다
        dp[value] = std::min(dp[value], cost);

        // 인구수 증가
        for(int j=value; j<=C+100; ++j)
        {
            dp[j] = std::min(dp[j], dp[j - value] + cost);
        }

    }

    for (int i = C; i <= C + 100; ++i) {
        answer = min(answer, dp[i]);
    }


    cout << answer;
    return 0;
}